src/sentry/scripts/digests/digests.lua

```diff
@@ -0,0 +1,318 @@
+    redis.call('EXPIRE', configuration:get_timeline_key(timeline_id), configuration.ttl)
+
+    local ready = add_timeline_to_schedule(configuration, timeline_id, timestamp, delay_increment, delay_maximum)
+
+    -- TODO: `tonumber` should happen upstream
+    local timeline_capacity = tonumber(timeline_capacity)
+    if timeline_capacity > 0 and math.random() < tonumber(truncation_chance) then
+        -- TODO: `tonumber` should happen upstream
+        truncate_timeline(configuration, timeline_id, timeline_capacity)
+    end
+
+    return ready
+end
+
+local function digest_timeline(configuration, timeline_id)
+    -- Check to ensure that the timeline is in the correct state.
+    if redis.call('ZSCORE', configuration:get_schedule_ready_key(), timeline_id) == false then
+        error('err(invalid_state): timeline is not in the ready state, cannot be digested')
+    end
+
+    local digest_key = configuration:get_timeline_digest_key(timeline_id)
+    local timeline_key = configuration:get_timeline_key(timeline_id)
+    if redis.call('EXISTS', digest_key) == 1 then
+        -- If the digest set already exists (possibly because we already tried
+        -- to send it and failed for some reason), merge any new data into it.
+        -- TODO: It might make sense to trim here to avoid returning capacity *
+        -- 2 if timeline was full when it was previously digested.
+        redis.call('ZUNIONSTORE', digest_key, 2, timeline_key, digest_key, 'AGGREGATE', 'MAX')
+        redis.call('DEL', timeline_key)
+        redis.call('EXPIRE', digest_key, configuration.ttl)
+    elseif redis.call('EXISTS', timeline_key) == 1 then
+        -- Otherwise, we can just move the timeline contents to the digest key.
+        redis.call('RENAME', timeline_key, digest_key)
+        redis.call('EXPIRE', digest_key, configuration.ttl)
+    end
+
+    local results = {}
+    local records = redis.call('ZREVRANGE', digest_key, 0, -1, 'WITHSCORES')
+    for key, score in zrange_scored_iterator(records) do
+        table.insert(results, {
+            key,
+            redis.call('GET', configuration:get_timeline_record_key(timeline_id, key)),
+            score
+        })
+    end
+
+    return results
+end
+
+local function close_digest(configuration, timeline_id, delay_minimum, ...)
+    local record_ids = {...}
+    local timeline_key = configuration:get_timeline_key(timeline_id)
+    local digest_key = configuration:get_timeline_digest_key(timeline_id)
+
+    if #record_ids > 0 then
+        redis.call('ZREM', digest_key, unpack(record_ids))
+        for _, record_id in ipairs(record_ids) do
+            -- TODO: This could technically be called as a variadic instead, if it mattered.
+            redis.call('DEL', configuration:get_timeline_record_key(timeline_id, record_id))
+        end
+    end
+
+    -- We always add to the ready set if we digested any number of records or
+    -- there are contents waiting to be delivered.
```

Comment:
I think this was another victim of the code changing and the comment not being updated, what you said is probably a clearer way to word it.

What I really meant was that if we didn't do anything (`record_ids` is empty) and if there isn't anything else left to do (there are no contents in the timeline or digest sorted sets), we can remove the timeline reference from both schedule sets.